### PR TITLE
python variables pane: don't hide `_` if it was changed by the user

### PR DESCRIPTION
**Pull request by austin3dickey**
Addresses #3366. This PR changes the Variables pane behavior in the Python extension so that it no longer hides the `_` variable if it was explicitly changed by the user.

An example: type `_ = 1` into the console. Before this change, `_` wouldn't show up. Now it does.

#### New Features

- N/A

#### Bug Fixes

- Allowed `_` to show up in the Python Variables pane if defined by the user. (#3366)

### QA Notes

Try messing around with the `_` variable; it should work like other variables in the Variables pane. Examples:
```py
_ = 1
_ = 2
_ = [1] * 1000
del _
from ibis import _
```
(the last one requires `pip install ibis-framework`)

@:console @:variables

### CODE REVIEW

File: extensions/positron-python/python_files/posit/positron/variables.py [hunk 1]
```diff
@@ -417,8 +412,17 @@ def _check_ns_subset(ns_subset, evaluated, are_different_func):
...
-        return self.kernel.shell.user_ns_hidden or {}
+    def _is_hidden(self, name: str, value: Any) -> bool:
+        """Is this variable a hidden kernel variable?.
+
+        Most of the time the answer is just whether it's in the kernel-hidden user namespace. But
+        the _ symbol is commonly overridden by users/packages. So we don't want to hide it if its
+        value is different from the value in the hidden namespace.
+        """
+        hidden = self.kernel.shell.user_ns_hidden or {}
+        if name == "_":
```
Comment (austin3dickey) [inline extensions/positron-python/python_files/posit/positron/variables.py:423]: Originally I was planning to take this fork for _any_ hidden variable. The hidden variable namespace includes (but isn't limited to) these variables:
<details>
  <summary>hidden variables</summary>

```
_ih
_oh
_dh
In
Out
get_ipython
exit
quit
open
help
_exit_code
__pydevd_ret_val_dict
__warningregistry__
__nonzero__
__name__
__doc__
__package__
__loader__
__spec__
__builtin__
__builtins__
_
__
___
_i
_ii
_iii
_i1
_i2
```
</details>

However, we started poking around at that change and realized that some hidden variables don't play well.
- `_exit_code` shows up when you run a `%pip install` magic that fails to install something, because at one point during that process, we start inspecting the variables and `_exit_code` happens to not match the hidden-namespace `_exit_code`. So this code thought the user had defined it.
- `__` was behaving strangely when I ran some tests that dealt with a complicated plot, for a similar reason.
- More?? There are probably more edge cases.

I believe these problems would have been fixable with more time and effort spent. But fixing those problems seems super low-value, and wasn't in the original scope of the issue anyway. So I'm proposing this PR just fixes the behavior of `_`. We can always revisit if someone needs another variable to show up.

**Comment by austin3dickey** [comment]
Just adding some tests now.

File: extensions/positron-python/python_files/posit/positron/tests/conftest.py [hunk 2]
```diff
@@ -75,7 +75,7 @@ def _prepare_shell(shell: PositronShell) -> None:
     #       manually add them to user_ns_hidden to replicate running in Positron.
     shell.user_ns_hidden.update(
         {
-            k: None
+            k: ""
```
Comment (austin3dickey) [inline extensions/positron-python/python_files/posit/positron/tests/conftest.py:78]: This is closer to what actually happens (and fixes a test for me)

File: extensions/positron-python/python_files/posit/positron/tests/test_variables.py
Comment (austin3dickey) [inline extensions/positron-python/python_files/posit/positron/tests/test_variables.py:1]: I went through this file and added a test case for the variable name `_` each time we do something with a variable. This was actually really valuable because it caught a lot of stuff.

File: extensions/positron-python/python_files/posit/positron/tests/test_variables.py [hunk 3]
```diff
@@ -348,34 +361,41 @@ async def test_clear(
...
-        json_rpc_notification("refresh", {"length": 0, "variables": [], "version": 0}),
+        json_rpc_notification("refresh", {"length": 1, "variables": [underscore], "version": 0}),
     ]
 
     # All user variables are removed
     assert "x" not in shell.user_ns
     assert "y" not in shell.user_ns
 
+    # ...except hidden variables, because %reset doesn't touch those
+    assert "_" in shell.user_ns
```
Comment (austin3dickey) [inline extensions/positron-python/python_files/posit/positron/tests/test_variables.py:372]: I learned that `%reset -s` (the [soft](https://ipython.readthedocs.io/en/stable/interactive/magics.html#magic-reset) reset) does not touch the underscore variable. So when we clear all variables, which calls this:

https://github.com/posit-dev/positron/blob/6e310bb5e7d9e2796f280c91000b9cbd55268724/extensions/positron-python/python_files/posit/positron/variables.py#L492

it doesn't remove the `_` from the variables pane. I decided to leave that behavior as-is instead of messing with the reset command. So now the variables pane simply reflects what's happened, in that everything gets deleted except `_`, even when it's been overridden by the user.

File: extensions/positron-python/python_files/posit/positron/tests/test_variables.py [hunk 4]
```diff
@@ -328,54 +340,65 @@ async def test_clear(
...
 
-    # All user variables are removed
-    assert "x" not in shell.user_ns
-    assert "y" not in shell.user_ns
+        # All user variables are removed
+        assert "x" not in shell.user_ns
+        assert "y" not in shell.user_ns
 
+        # ...except hidden variables, because %reset -s doesn't touch those
+        assert "_" in shell.user_ns
```
Comment (austin3dickey) [inline extensions/positron-python/python_files/posit/positron/tests/test_variables.py:375]: I learned that `%reset -s` (the [soft](https://ipython.readthedocs.io/en/stable/interactive/magics.html#magic-reset) reset) does not touch the underscore variable. So when we clear all variables, which calls this:

https://github.com/posit-dev/positron/blob/6e310bb5e7d9e2796f280c91000b9cbd55268724/extensions/positron-python/python_files/posit/positron/variables.py#L492

it doesn't remove the `_` from the variables pane. I decided to leave that behavior as-is instead of messing with the reset command. So now the variables pane simply reflects what's happened, in that everything gets deleted except `_`, even when it's been overridden by the user.

File: extensions/positron-python/python_files/posit/positron/variables.py (hunk 1 above)
Comment (isabelizimm) [reply extensions/positron-python/python_files/posit/positron/variables.py:423]: I don't see `__` and `_exit_code` often (ever?) in the wild, but `_` is pretty common. I agree this is the right level of fix for now!

File: extensions/positron-python/python_files/posit/positron/tests/test_variables.py [hunk 5]
```diff
@@ -348,34 +361,41 @@ async def test_clear(
...
         ),
-        json_rpc_notification("refresh", {"length": 0, "variables": [], "version": 0}),
+        json_rpc_notification("refresh", {"length": 1, "variables": [underscore], "version": 0}),
     ]
 
     # All user variables are removed
     assert "x" not in shell.user_ns
     assert "y" not in shell.user_ns
 
+    # ...except hidden variables, because %reset doesn't touch those
```
Comment (isabelizimm) [inline extensions/positron-python/python_files/posit/positron/tests/test_variables.py]: woah! TIL 🤯

**Review by isabelizimm — APPROVED** [review]
works as anticipated! 🚀

File: extensions/positron-python/python_files/posit/positron/tests/test_variables.py [hunk 6]
```diff
@@ -97,41 +97,45 @@ def test_comm_open(kernel: PositronIPyKernel) -> None:
             ],
         ),
         # Nested mutable types
-        ("", ["x = [{}]", "x[0]['a'] = 0"]),
-        ("", ["x = {'a': []}", "x['a'].append(0)"]),
+        ("", [" = [{}]", "[0]['a'] = 0"]),
+        ("", [" = {'a': []}", "['a'].append(0)"]),
     ],
 )
+@pytest.mark.parametrize("varname", ["x", "_"])
```
Comment (seeM) [inline extensions/positron-python/python_files/posit/positron/tests/test_variables.py:104]: Woah, I did not know that you can chain `parametrize`, very useful!

File: extensions/positron-python/python_files/posit/positron/tests/test_variables.py
Comment (seeM) [reply extensions/positron-python/python_files/posit/positron/tests/test_variables.py:1]: It's useful to have `varname` parametrized as well instead of `"x"` hardcoded everywhere.

**Review by seeM — APPROVED** [review]
Code changes LGTM, great to see the tests!

File: extensions/positron-python/python_files/posit/positron/tests/test_variables.py (hunk 6 above)
Comment (austin3dickey) [reply extensions/positron-python/python_files/posit/positron/tests/test_variables.py:104]: yeah at my last job I was guilty of writing a test with like 6 different parametrizations resulting in like 3500 cases 😂 coverage was really important!!
